### PR TITLE
litep2p/peerset: Do not disconnect all peers on `SetReservedPeers` command

### PR DESCRIPTION
**Pull request by lexnv**
Previously, when receiving the `SetReservedPeers { reserved }` all peers not in the `reserved` set were removed.

This is incorrect, the intention of `SetReservedPeers` is to change the active set of reserved peers and disconnect previously reserved peers not in the new set. 

While at it, have added a few other improvements to make the peerset more robust:
- `SetReservedPeers`: does not disconnect all peers
- `SetReservedPeers`: if a reserved peer is no longer reserved, the peerset tries to move the peers to the regular set if the slots allow this move. This ensures the (now regular) peer counts towards slot allocation. 
- every 1 seconds: If we don't have enough connect peers, add the reserved peers to the list that the peerstore ignores. Reserved peers are already connected and the peerstore might return otherwise a reserved peer 

### Next Steps

- [x] More testing

cc @paritytech/networking

### CODE REVIEW

File: substrate/client/network/src/litep2p/shim/notification/peerset.rs [hunk 1]
```diff
@@ -960,39 +974,64 @@ impl Stream for Peerset {
...
-						.filter_map(|(peer, _)| (!peers.contains(peer)).then_some(*peer))
-						.collect::<HashSet<_>>();
+					// collect all *reserved* peers who are not in the new reserved set
+					let reserved_peers_maybe_remove =
+						self.reserved_peers.difference(&peers).cloned().collect::<HashSet<_>>();
 
 					self.reserved_peers = peers;
 
-					let peers = peers_to_remove
+					let peers = reserved_peers_maybe_remove
```
Comment (dmitry-markin) [inline substrate/client/network/src/litep2p/shim/notification/peerset.rs:991]:
```suggestion
					let peers_to_remove = reserved_peers_maybe_remove
```

**Review by dmitry-markin — APPROVED** [review]
Looks good!

**Review by michalkucharczyk — APPROVED** [review]
Not really my domain, made review for learning purposes. From what I can say looks good. Left some nitpicks.

File: substrate/client/network/src/litep2p/shim/notification/peerset.rs [hunk 2]
```diff
@@ -960,39 +974,64 @@ impl Stream for Peerset {
...
+									// The direction contains a `Reserved::Yes` flag, because this
+									// is a reserve peer that we want to close.
+									// The `Reserved::Yes` ensures we don't adjust the slot count
+									// when the substream is closed.
+
+									let disconnect = self.reserved_only ||
+										match direction {
+											Direction::Inbound(_) => self.num_in >= self.max_in,
+											Direction::Outbound(_) => self.num_out >= self.max_out,
+										};
```
[the remark this answers is not in the document]
Comment (bkchr) [reply substrate/client/network/src/litep2p/shim/notification/peerset.rs:1005]: This piece of code was replicated three times. Could be moved into one function that takes a `disconnect_state` and a `keep_state`.

**Review by bkchr — APPROVED** [review]

File: substrate/client/network/src/litep2p/shim/notification/tests/peerset.rs [hunk 3]
```diff
@@ -885,7 +1023,100 @@ async fn set_reserved_peers_but_available_slots() {
...
+				assert_eq!(
+					peerset.peers().get(&peer),
+					Some(&PeerState::Opening { direction: Direction::Outbound(Reserved::Yes) })
+				);
+			}
+		},
+		event => panic!("invalid event: {event:?}"),
+	}
+
+	// verify all three peers are counted as outbound peers and they don't count towards
```
[the remark this answers is not in the document]
Comment (lexnv) [reply substrate/client/network/src/litep2p/shim/notification/tests/peerset.rs:1077]: I think that was a mistake, thanks for pointing that out 🙏

**Pull request merged**